### PR TITLE
API: Make most arguments for read_html and read_json keyword-ony

### PR DESCRIPTION
**Pull request by alexitkes**
As mentioned in #27544 it is better to use keyword-only arguments for functions with too many arguments. A deprecation warning will be displayed if `read_html` get more than 2 positional arguments (`io` and `match`) or `read_json` get more than 1 positional argument (`path_or_buf`).

- [x] tests added / passed
   Three groups of tests are actually needed
   - [x] Tests dof the `deprecate_nonkeyword_arguments` decoratorr
   - [x] Check whether the `read_json` function emits a `FutureWarning` whenever necessary and does not emit it whenever not.
   - [x] Check whether the `read_html` function emits a `FutureWarning` whenever necessary and does not emit it whenever not.
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
We need to deprecate the old behavior first. As written, this is an api breaking change.

And we shouldn’t do this until 1.0 or later.

**Comment by jreback** [comment]
> We need to deprecate the old behavior first. As written, this is an api breaking change.
> 
> And we shouldn’t do this until 1.0 or later.

and how exactly would you deprecate ?

**Comment by TomAugspurger** [comment]
Should be able to use a decorator to inspect the calm, right? And if any positional or keyword args are passed in args then we warn?

**Comment by alexitkes** [comment]
Thank you. I also wanted to propose somewhat like this.

```
def Deprecated(*, n_args: int):
    def _deprecated(f):
        def _f(*args, **kwargs):
            if len(args) > n_args:
                warnings.warn("%s will only accept %i positional arguments in version X.Y.Z" % (f.__name__,n_args))
            return f(*args, **kwargs)
        return _f
    return _deprecated
```

Or do such a decorator already exist?

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -7,6 +7,18 @@
 What's new in 0.25.1 (July XX, 2019)
 ------------------------------------
 
+.. warning::
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.1.rst]: This should go to 1.0. We don't add deprecations in point releases.

There should be "Deprecations" section.

Don't need the warning boxes.

Do need a link back to the issue number with :issue:

File: pandas/io/html.py [hunk 2]
```diff
@@ -920,6 +921,7 @@ def _parse(flavor, io, match, attrs, encoding, displayed_only, **kwargs):
     return ret
 
 
+@deprecate_nonkeyword_arguments(version="1.0", allowed_args=["io", "match"])
```
Comment (TomAugspurger) [inline pandas/io/html.py]: Why are we allowing `match` as a positional argument?

I'd like not to repeat the arguments here. That seems fragile. Is it possible to have the decorator inspect the signature and warn on any positional_or_keyword arg passed positionally?

**Review by TomAugspurger — COMMENTED** [review]
Can you add tests to read_html / read_json to ensure that warnings are raised there?

**Comment by TomAugspurger** [comment]
Lifting https://github.com/pandas-dev/pandas/pull/27573#discussion_r308181531 to the top level

> I'd like not to repeat the arguments here. That seems fragile. Is it possible to have the decorator inspect the signature and warn on any positional_or_keyword arg passed positionally?

What do you think about this @alexitkes?

**Comment by alexitkes** [comment]
@TomAugspurger Thank you for review. Yes, writing argument list twice will be quite annoying, but I will need some time to find a way to avoid it, and I would like to try. However, I think it would be better to preserve the `allowed_args` argument for the decorator and default it to list of required arguments if None given.

**Comment by TomAugspurger** [comment]
> However, I think it would be better to preserve the allowed_args argument for the decorator and default it to list of required arguments if None given.

Why do you think we'll need that? That sounds more complicated than just detecting kwargs passed positionally. If we aren't going to need it then I'd rather not implement it.

**Comment by alexitkes** [comment]
@TomAugspurger I have just written such a decorator and it does not look too complicated. However it is my first experience with `inspect` module so possibly more tests are needed though some tests are written and they all run successfully.

It will not be difficult to remove `allowed_args` argument and always set it to list of required arguments, if you think eventually all arguments having the default values should be keyword-only arguments. Should I do it?

**Comment by TomAugspurger** [comment]
I'm not sure without looking more closely. My thought is that all keyword
arguments being keyword-only is easier to explain to users.

**Comment by alexitkes** [comment]
> I'm not sure without looking more closely. My thought is that all keyword
> arguments being keyword-only is easier to explain to users.

I have almost forgotten one thing - the `path_or_buf` argument for `read_json` has the default value `None`. So if we decorate the function in that way all arguments for `read_json` will be keyword-only. But I think most users are used to writing `read_json(filename)` and will not like to write `read_json(path_or_buf=filename)`. For this reason I still think there should be a way to pass the list of arguments allowed to be positional to the decorator. However, the default decorator behavior (optional arguments are keyword only) should be used wherever there is no really significant reason to change it. The read_html function is decorated in this way.

File: doc/source/whatsnew/v0.25.1.rst (hunk 1 above)
Comment (alexitkes) [reply doc/source/whatsnew/v0.25.1.rst]: Thank you, moved it to v1.0.0.rst

**Comment by TomAugspurger** [comment]
> I have almost forgotten one thing - the path_or_buf argument for read_json has the default value None. So if we decorate the function in that way all arguments for read_json will be keyword-only.

Oh sorry. In that case we'll definitely need to have the ability to have some keyword-or-positional arguments.

**Comment by alexitkes** [comment]
Well, it seems to work fine though I am a bit worried because it's my first experience with `inspect` module. I also don't know the proper number of version in which it will become impossible to use positionl arguments (currently written 1.1).  What should it be?

File: pandas/io/html.py [hunk 3]
```diff
@@ -920,6 +921,7 @@ def _parse(flavor, io, match, attrs, encoding, displayed_only, **kwargs):
     return ret
 
 
+@deprecate_nonkeyword_arguments(version="1.1")
```
Comment (TomAugspurger) [inline pandas/io/html.py]: Usually we go at least 3 releases, so this would be 1.4

File: pandas/tests/io/json/test_deprecated_kwargs.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def test_deprecated_kwargs():
+    """
+    Check whether FutureWarning is produced if `orient`
+    argument passed as positional.
+    """
+    df = pd.DataFrame({"A": [2, 4, 6], "B": [3, 6, 9]}, index=[0, 1, 2])
+    with tm.assert_produces_warning(FutureWarning):
+        assert_frame_equal(df, read_json(df.to_json(orient="split"), "split"))
```
Comment (TomAugspurger) [inline pandas/tests/io/json/test_deprecated_kwargs.py]: Can you minimize the amount of code in the context manager? I think just `read_json(buf, "split")`.

Also, do we need to test all three of these? What differs between them?

File: pandas/tests/io/json/test_deprecated_kwargs.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        assert_frame_equal(df, read_json(df.to_json(orient="index"), "index"))
+
+
+def test_good_kwargs():
+    """
+    Check whether FutureWarning is not produced if `orient`
+    argument passed as keyword.
+    """
+    df = pd.DataFrame({"A": [2, 4, 6], "B": [3, 6, 9]}, index=[0, 1, 2])
+    assert_frame_equal(df, read_json(df.to_json(orient="split"), orient="split"))
```
Comment (TomAugspurger) [inline pandas/tests/io/json/test_deprecated_kwargs.py]: I think use `tm.assert_produces_warning(None)`

File: pandas/util/_decorators.py [hunk 6]
```diff
@@ -212,6 +212,61 @@ def wrapper(*args, **kwargs):
...
+    """
+    Decorator to deprecate a use of non-keyword arguments of a function.
+
+    Parameters
+    ----------
+    version : str
+        The version in which positional arguments will become
+        keyword-only.
+
+    allowed_args : list, int or None, default=None
```
Comment (TomAugspurger) [inline pandas/util/_decorators.py]: `or None, default=None` -> optional.

File: pandas/util/_decorators.py [hunk 7]
```diff
@@ -212,6 +212,61 @@ def wrapper(*args, **kwargs):
...
+            spec = inspect.getfullargspec(func)
+            allow_args = spec.args[: -len(spec.defaults)]
+        else:
+            allow_args = allowed_args
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            if isinstance(allow_args, int) and len(args) > allow_args:
+                msg = (
+                    "After version %s all arguments of %s "
```
Comment (TomAugspurger) [inline pandas/util/_decorators.py]: version -> pandas version.

Use `{}`-style string templates.

File: pandas/tests/util/test_deprecate_nonkeyword_arguments.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    """
+    return a + b + c + d
+
+
+def test_one_arguments():
+    """
+    Check whether no future warning is produced if one
+    positional argument given.
+    """
+    assert f(19) == 19
```
Comment (TomAugspurger) [inline pandas/tests/util/test_deprecate_nonkeyword_arguments.py]: Need to assert that there's no warnings where relevant. Our CI sets things up so it's checked, but we shouldn't rely on that.

File: pandas/io/html.py (hunk 3 above)
Comment (alexitkes) [reply pandas/io/html.py]: Thank you, changed it.

File: pandas/tests/util/test_deprecate_nonkeyword_arguments.py (hunk 8 above)
Comment (alexitkes) [reply pandas/tests/util/test_deprecate_nonkeyword_arguments.py]: Ok, I used `tm.assert_produces_warning(None)` just like in pandas/tests/io/json/test_deprecated_kwargs.py. Is it the proper way?

Comment (TomAugspurger) [reply pandas/tests/util/test_deprecate_nonkeyword_arguments.py]: I think so.

File: pandas/util/_decorators.py (hunk 6 above)
Comment (alexitkes) [reply pandas/util/_decorators.py]: OK

**Comment by TomAugspurger** [comment]
@WillAyd @jreback @jorisvandenbossche do you have thoughts on this (the overall goal, the implementation seems fine).

* Do we expect to expand this to more readers? Should these be done in the same release?
* What about other places in the library? Same release?
* What's a good general policy on whether an argument should be keyword only?

File: pandas/util/_decorators.py (hunk 7 above)
Comment (alexitkes) [reply pandas/util/_decorators.py]: Ok, done it. I also added a separate function to form a better looking warning message depending on the number of arguments, but so far I only tested it manually. Is there a proper way to test how exactly does the produced warning look?

File: pandas/io/html.py [hunk 9]
```diff
@@ -920,6 +921,7 @@ def _parse(flavor, io, match, attrs, encoding, displayed_only, **kwargs):
     return ret
 
 
+@deprecate_nonkeyword_arguments(version="1.4")
```
Comment (WillAyd) [inline pandas/io/html.py]: Does `io` need to be in `allowed_args` here?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
@TomAugspurger all good questions to which I don't have a definitive answer. Doing all in one release would be ideal though not sure how feasible that really is.

Knowing there's going to be exceptions to anything we do we could consider something simple like saying "any functions with more than N arguments will be keyword only" where N could be 5 or less.

File: pandas/io/html.py (hunk 9 above)
Comment (alexitkes) [reply pandas/io/html.py]: This is not needed. If no `allowed_args` given, it defaults to the list of arguments not having the default value.

File: pandas/util/_decorators.py [hunk 10]
```diff
@@ -212,6 +212,102 @@ def wrapper(*args, **kwargs):
...
+        integer, this is the number of positional arguments
+        that will stay positional. In case of None value,
+        defaults to list of all arguments not having the
+        default value.
+
+    stacklevel : int, default=2
+        The stack level for warnings.warn
+    """
+
+    def format_argument_list(allow_args):
```
[the remark this answers is not in the document]
Comment (alexitkes) [reply pandas/util/_decorators.py]: Ok, done it.

File: pandas/tests/io/test_html.py [hunk 11]
```diff
@@ -168,81 +220,122 @@ def test_banklist_no_match(self):
...
 
     def test_skiprows_invalid(self):
         with pytest.raises(TypeError, match=("is not a valid type for skipping rows")):
-            self.read_html(self.spam_data, ".*Water.*", skiprows="asdf")
+            self.read_html(self.spam_data, match=".*Water.*", skiprows="asdf")
 
     def test_index(self):
-        df1 = self.read_html(self.spam_data, ".*Water.*", index_col=0)
-        df2 = self.read_html(self.spam_data, "Unit", index_col=0)
+        with tm.assert_produces_warning(FutureWarning):
```
[the remark this answers is not in the document]
Comment (alexitkes) [reply pandas/tests/io/test_html.py]: Well, such amount of duplicated tests looks quite ugly, so I removed all tests with deprecated behavior except for one. Would it be enough?

File: pandas/tests/util/test_deprecate_nonkeyword_arguments.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+def f(a, b=0, c=0, d=0):
+    """
+    Sum of one to four numbers, the two last arguments
+    should be given as keyword arguments only
+    """
+    return a + b + c + d
+
+
+def test_one_argument():
+    """
```
[the remark this answers is not in the document]
Comment (alexitkes) [reply pandas/tests/util/test_deprecate_nonkeyword_arguments.py]: OK.

**Comment by TomAugspurger** [comment]
This seems to have stalled a bit.

Are we comfortable going down this API path? Any other changes needed here? (cc @gfyoung, @WillAyd when you're back).

**Comment by gfyoung** [comment]
I'm relatively comfortable with this path (it was discussed a good amount in the original issue).  However, it would be great to see some tests where we introspect on the warning message, given the amount of logic happening there.

File: pandas/util/_decorators.py [hunk 13]
```diff
@@ -212,6 +212,105 @@ def wrapper(*args, **kwargs):
     return _deprecate_kwarg
 
 
+def _format_argument_list(allow_args: Union[List[str], int]):
```
Comment (WillAyd) [inline pandas/util/_decorators.py]: Instead of cramming allowed_args into one variable can we have separate parameters for positional and keyword argument allowance? I think a more natural fit that way

**Review by WillAyd — CHANGES_REQUESTED** [review]
Personally having second thoughts on general approach. Having a hard time thinking through why we would want to make exceptions on a per function basis, so some of the allowances here may be overkill.

Still wondering if it's just not easier to say something like "a maximum of N positional arguments will be allowed in pandas; everything else is keyword only" (maybe N=5) as a general rule of thumb and messaging that somehow. Would I think be a lot easier to implement and maintain

**Comment by TomAugspurger** [comment]
Hmm I disagree with that. The deprecation seems pretty clean to me, just a
single decorator.

**Comment by alexitkes** [comment]
Thank you for reviews.

@gfyoung Yes, some introspection of the warning messages would be good. Unfortunately, I can't find a way to do it within `pandas.util.testing` module. Is there a proper way to examine warning messages, or it only can be done by using `warnings.catch_warnings` directly? In letter case I will still need some time to learn...

> Instead of cramming allowed_args into one variable can we have separate parameters for positional and keyword argument allowance? I think a more natural fit that way

@WillAyd Sorry, but I am afraid I don't understand what exactly you mean, possibly due to not being a native English-speaker. Can you give an example of how should a good warning message look?

**Comment by gfyoung** [comment]
@alexitkes : I would attempt to do it via `warnings.catch_warnings`.

**Comment by alexitkes** [comment]
Added some tests with warning message analysis to `test_deprecate_nonkeyword_arguments.py`.

File: pandas/tests/io/json/test_deprecated_kwargs.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+import pandas as pd
+import pandas.util.testing as tm
+from pandas.util.testing import assert_frame_equal
+
+from pandas.io.json import read_json
+
+
+def test_deprecated_kwargs():
+    """
```
Comment (WillAyd) [inline pandas/tests/io/json/test_deprecated_kwargs.py]: We don't typically use docstrings for tests so OK to remove here and below

File: pandas/util/_decorators.py [hunk 15]
```diff
@@ -226,6 +226,105 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
...
+        args = ", ".join(["'" + x + "'" for x in allow_args[:-1]])
+        return " except for the arguments {args} and '{last}'".format(
+            args=args, last=last
+        )
+
+
+def deprecate_nonkeyword_arguments(
+    version: str,
+    allowed_args: Optional[Union[List[str], int]] = None,
+    stacklevel: int = 2,
```
Comment (WillAyd) [inline pandas/util/_decorators.py]: Hmm why is stack level part of the signature here?

Comment (alexitkes) [reply pandas/util/_decorators.py]: I just done it like it was in the `deprecate_kwarg` decorator serving a close purpose.

Note that the stack level argument of `deprecate_kwarg` function currently seems to be never used, but if there is a reason to have it, it may be valueable here too.

File: pandas/tests/io/json/test_deprecated_kwargs.py (hunk 14 above)
Comment (alexitkes) [reply pandas/tests/io/json/test_deprecated_kwargs.py]: OK.

**Comment by WillAyd** [comment]
@pandas-dev/pandas-core do we have strong thoughts on this one? Want to be cognizant of @alexitkes time if we aren't aligned

FWIW I'm not sure this is worth the effort

**Comment by jreback** [comment]
i would just change this and call it a day
depreciation is not worth the effort here

**Comment by WillAyd** [comment]
I also agree. I think everything after the filepath_or_buf argument should be keyword only if we do this at all

There will be some friction but its very easy to recover from and sets a consistent go-forward expectation. 1.0 is the time to do something like that too

**Comment by TomAugspurger** [comment]
Is the issue the complication of the deprecation, or disagreement on what should be keyword only.

I’m -1 on a breaking change. I don’t think the deprecation core complicates things much. 

No strong thoughts on the issue with what becomes keyword only. Is the fact that the builtin open method doesn’t do keyword only args helpful in deciding?

**Comment by Sankareswari-Ramamoorthy** [comment]
\@Sankareswari-Ramamoorthy

File: doc/source/whatsnew/v1.0.0.rst [hunk 16]
```diff
@@ -169,6 +169,15 @@ Deprecations
 - ``Index.set_value`` has been deprecated. For a given index ``idx``, array ``arr``,
   value in ``idx`` of ``idx_val`` and a new value of ``val``, ``idx.set_value(arr, idx_val, val)``
   is equivalent to ``arr[idx.get_loc(idx_val)] = val``, which should be used instead (:issue:`28621`).
+
+- Passing any arguments but `io` to  :func:`read_html` as positional
+  arguments is deprecated since version 1.0. All other arguments should
+  be given as keyword arguments (:issue:`27573`).
+
+- Passing any arguments but `path_or_buf` to  :func:`read_json` as positional
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can you create an issue to enforce this deprecation in the future

**Review by jreback — APPROVED** [review]
looks good.

**Comment by jreback** [comment]
can you merge master; ping on green.

File: pandas/tests/io/json/test_deprecated_kwargs.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
+"""
+Tests for the deprecated keyword arguments for `read_json`.
+"""
+
+import pandas as pd
+import pandas.util.testing as tm
+from pandas.util.testing import assert_frame_equal
```
Comment (jbrockmendel) [inline pandas/tests/io/json/test_deprecated_kwargs.py]: the linter wants you to use tm.assert_frame_equal instead of importing it

**Comment by jbrockmendel** [comment]
We have something in plotting._core to warn about positional arguments, maybe should use this new decorator there (separate PR)

**Comment by jbrockmendel** [comment]
@alexitkes can you rebase and handle the last couple comments?  this is going to be really nice tom get done

File: pandas/tests/io/json/test_deprecated_kwargs.py (hunk 17 above)
Comment (alexitkes) [reply pandas/tests/io/json/test_deprecated_kwargs.py]: OK.

File: doc/source/whatsnew/v1.0.0.rst [hunk 18]
```diff
@@ -523,6 +523,12 @@ Deprecations
   it is recommended to use ``json_normalize`` as :func:`pandas.json_normalize` instead (:issue:`27586`).
 - :meth:`DataFrame.to_stata`, :meth:`DataFrame.to_feather`, and :meth:`DataFrame.to_parquet` argument "fname" is deprecated, use "path" instead (:issue:`23574`)
 - The deprecated internal attributes ``_start``, ``_stop`` and ``_step`` of :class:`RangeIndex` now raise a ``FutureWarning`` instead of a ``DeprecationWarning`` (:issue:`26581`)
+- Passing any arguments but `io` to  :func:`read_html` as positional
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]: Can you move note to 1.1 at this point?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
looks reasonable, can you rebase

File: doc/source/whatsnew/v1.0.0.rst (hunk 18 above)
Comment (alexitkes) [reply doc/source/whatsnew/v1.0.0.rst]: OK

**Comment by alexitkes** [comment]
Just rebased to master. The read_json function has now both deprecate_kwarg and deprecate_nonkeyword_args decorators and both messages are displayed if necessary.

Note in whatsnew was moved to v1.1, but nonkeyword arguments are still promised to be removed since v1.4.

File: doc/source/whatsnew/v1.1.0.rst [hunk 19]
```diff
@@ -64,8 +64,12 @@ Other API changes
 Deprecations
 ~~~~~~~~~~~~
 
--
--
+- Passing any arguments but `io` to  :func:`read_html` as positional
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.1.0.rst]: I would maybe say here "any arguments but the first" (since `io` doesn't say much if you don't know the exact names of the signature)

File: doc/source/whatsnew/v1.1.0.rst [hunk 20]
```diff
@@ -64,8 +64,12 @@ Other API changes
...
+  arguments is deprecated since version 1.0. All other arguments should
+  be given as keyword arguments (:issue:`27573`).
+- Passing any arguments but `path_or_buf` to  :func:`read_json` as positional
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.1.0.rst]: I think you can combine both entries in one

File: pandas/io/html.py [hunk 21]
```diff
@@ -922,6 +923,7 @@ def _parse(flavor, io, match, attrs, encoding, displayed_only, **kwargs):
...
```
Comment (jorisvandenbossche) [inline pandas/io/html.py]:
```suggestion
@deprecate_nonkeyword_arguments(version="2.0")
```

Following the new versioning policy, this deprecation will be enforced in 2.0 (https://dev.pandas.io/docs/development/policies.html)

File: pandas/io/json/_json.py [hunk 22]
```diff
@@ -347,6 +347,9 @@ def _write(
 
 
 @deprecate_kwarg(old_arg_name="numpy", new_arg_name=None)
+@deprecate_nonkeyword_arguments(
+    version="1.4", allowed_args=["path_or_buf"], stacklevel=3
```
Comment (jorisvandenbossche) [inline pandas/io/json/_json.py]:
```suggestion
    version="2.0", allowed_args=["path_or_buf"], stacklevel=3
```

Comment (jorisvandenbossche) [inline pandas/io/json/_json.py]: The reason for the stacklevel=3 is because there is another decorator around it?

File: pandas/tests/io/test_html.py [hunk 23]
```diff
@@ -168,81 +185,88 @@ def test_banklist_no_match(self):
...
 
     def test_skiprows_range(self):
-        df1 = self.read_html(self.spam_data, ".*Water.*", skiprows=range(2))[0]
-        df2 = self.read_html(self.spam_data, "Unit", skiprows=range(2))[0]
+        df1 = self.read_html(self.spam_data, match=".*Water.*", skiprows=range(2))
+        df2 = self.read_html(self.spam_data, match="Unit", skiprows=range(2))
+
+        assert_framelist_equal(df1, df2)
+
+    def test_skiprows_range_single_frame(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/io/test_html.py]: Is this a new test?

File: pandas/util/_decorators.py [hunk 24]
```diff
@@ -218,6 +218,105 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
...
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            arguments = _format_argument_list(allow_args)
+            if isinstance(allow_args, (list, tuple)):
+                num_allow_args = len(allow_args)
+            else:
+                num_allow_args = allow_args
+            if len(args) > num_allow_args:
+                msg = (
+                    "After Pandas version {version} all arguments of {funcname}"
```
Comment (jorisvandenbossche) [inline pandas/util/_decorators.py]:
```suggestion
                    "After pandas version {version} all arguments of {funcname}"
```

(we generally use lower case in the docs)

File: pandas/util/_decorators.py [hunk 25]
```diff
@@ -218,6 +218,105 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
...
+                    "{except_args} will be keyword-only"
```
Comment (jorisvandenbossche) [inline pandas/util/_decorators.py]: I would maybe rephrase the message a little bit. Eg include something like "Passing positional arguments is deprecated" (and then continue with saying that all arguments except .. will be keyword only)

Also, should "after pandas version xx" be "starting with pandas version xx" ?

File: pandas/io/html.py (hunk 21 above)
Comment (jorisvandenbossche) [inline pandas/io/html.py]:
```suggestion
@deprecate_nonkeyword_arguments(version="1.4", allowed_args=["io"])
```

? 
I know it is not needed here (since it is inferred from the signature), but I think the deprecation message is better when this is specified (saying "the first").

**Review by jorisvandenbossche — COMMENTED** [review]
@alexitkes thanks for keeping pushing this PR. I added a bunch of comments, mostly cosmetic details on wording.

Also, you seem to have changed the permissions of the v1.0.0 file, can you try to revert that?

File: pandas/io/json/_json.py (hunk 22 above)
Comment (alexitkes) [reply pandas/io/json/_json.py]: Yes, it's looking like greater stack level is needed in case of multiple decorators.

File: pandas/tests/io/test_html.py (hunk 23 above)
Comment (alexitkes) [reply pandas/tests/io/test_html.py]: This test was changed while this PR in progress. I thought it would be safer to preserve both the new test and the test from my branch.

File: pandas/util/_decorators.py (hunk 25 above)
Comment (alexitkes) [reply pandas/util/_decorators.py]: OK.

**Comment by alexitkes** [comment]
> Also, you seem to have changed the permissions of the v1.0.0 file, can you try to revert that?

What permissions should it have? On my system, all whatsnew files including this one have `-rw-r--r--` permissions.

File: pandas/util/_decorators.py (hunk 25 above)
Comment (alexitkes) [reply pandas/util/_decorators.py]: OK

File: pandas/io/html.py (hunk 21 above)
Comment (alexitkes) [reply pandas/io/html.py]: Hmm, that would be reasonable if keep whatsnew file unchanged.
https://github.com/pandas-dev/pandas/pull/27573#discussion_r370082289

I think it's better to say either "all arguments buy `io`" or "all arguments but the first one" in both places.

**Comment by jorisvandenbossche** [comment]
> What permissions should it have?

You can see in the diff what it had originally: https://github.com/pandas-dev/pandas/pull/27573/files

**Comment by jorisvandenbossche** [comment]
Instead of changing the permissions specifically, I suppose you can also restore the state of that specific file to that of the master branch (which should indirectly also fix the permission change). I *think* something like `git checkout master -- doc/source/whatsnew/v1.0.0.rst` should do that, but didn't check (https://stackoverflow.com/questions/2364147/how-to-get-just-one-file-from-another-branch/2364223)

**Comment by alexitkes** [comment]
@jorisvandenbossche Yes, I see - it had executable flag. I reverted it back, but still don't think it should have executable permission because it does not contain executable header anyway. I also doubt about executable flag for core/indexing.py that also does not contain `#! command` header. Maybe their permissions should be changed in a separate PR?

**Comment by jorisvandenbossche** [comment]
Ah, indeed, `v1.0.0.rst` is the outlier, all the other files have properly no executable flag set.

File: pandas/tests/io/test_html.py (hunk 23 above)
Comment (jreback) [reply pandas/tests/io/test_html.py]: this looks almost exactly the same, ok i guess.

File: pandas/util/_decorators.py (hunk 13 above)
Comment (jreback) [reply pandas/util/_decorators.py]: can you do this?

**Review by jreback — CHANGES_REQUESTED** [review]
looks really good a few comments

File: pandas/tests/io/test_html.py (hunk 23 above)
Comment (alexitkes) [reply pandas/tests/io/test_html.py]: Well... so I remove the second one. Right?

File: pandas/util/_decorators.py (hunk 13 above)
Comment (alexitkes) [reply pandas/util/_decorators.py]: I am not sure I understand you right (not a native English speaker).

Does it mean, for example,
```
_format_argument_list(3, [x, y])
```

should return "except for the arguments x, y and the first 3 positional arguments"? Such warnings should never be produced by `deprecate_nonkeyword_arguments`, but may make this function useful to format some other warning messages. Is it right?

**Comment by alexitkes** [comment]
@jreback @WillAyd 

I am not sure I understand these replies right:
https://github.com/pandas-dev/pandas/pull/27573/files#r317349159
https://github.com/pandas-dev/pandas/pull/27573#discussion_r370967374

Did you mean, for example, that

```
_format_argument_list(3, [x, y])
```

should return "except for the arguments x, y and the first 3 positional arguments"? Such warnings should never be produced by deprecate_nonkeyword_arguments, but may make this function useful to format some other warning messages. Is it right?

File: doc/source/whatsnew/v1.1.0.rst (hunk 19 above)
Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst]: can you update this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls merge master

**Comment by WillAyd** [comment]
@alexitkes can you fix up the merge conflict?

**Comment by alexitkes** [comment]
@WillAyd OK, done it.

**Review by WillAyd — APPROVED** [review]
I would be OK with moving forward here - @jreback or @TomAugspurger ?

**Comment by TomAugspurger** [comment]
Do we have a general policy for which arguments are keyword only?

**Comment by jreback** [comment]
lgtm. on anything with lots of args (pretty much read_*), happy to only have the first arg positional (path_or_buf).

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good, @TomAugspurger ok here?

**Comment by TomAugspurger** [comment]
Yeah

**Comment by jreback** [comment]
thanks!

**Comment by jorisvandenbossche** [comment]
Thanks @alexitkes !
